### PR TITLE
Bump Gradle to version 8.0

### PR DESCRIPTION
**Pull request by snazy**
Bumps Gradle version to 8.0, adopts the build scripts. No functional change.

### CODE REVIEW

File: devtools/gradle/gradle/wrapper/gradle-wrapper.properties [hunk 1]
```diff
@@ -1,5 +1,7 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-7.5.1-all.zip
+distributionSha256Sum=f30b29580fe11719087d698da23f3b0f0d04031d8995f7dd8275a31f7674dc01
+distributionUrl=https\://services.gradle.org/distributions/gradle-8.0-all.zip
+networkTimeout=10000
```
Comment (snazy) [inline devtools/gradle/gradle/wrapper/gradle-wrapper.properties:5]: `networkTimeout` is added when bumping the wrapper version using `./gradlew wrapper ...` - it was always kinda recommended to set it.

File: devtools/gradle/gradle-application-plugin/build.gradle [hunk 2]
```diff
@@ -24,14 +24,11 @@ gradlePlugin {
             implementationClass = 'io.quarkus.gradle.QuarkusPlugin'
             displayName = 'Quarkus Plugin'
             description = 'Builds a Quarkus application, and provides helpers to launch dev-mode, the Quarkus CLI, building of native images'
+            tags.addAll(['quarkus', 'quarkusio', 'graalvm'])
         }
     }
-}
-
-pluginBundle {
```
Comment (snazy) [inline devtools/gradle/gradle-application-plugin/build.gradle:31]: `pluginBundle` is gone now.

**Comment by famod** [comment]
Comparing to #29508 (which we had to close because of OOM issues), I think a few bits are missing here.

/cc @geoand @glefloch

**Comment by snazy** [comment]
> I think a few bits are missing here

Added the additional changes from #29508 that I could identify.

**Comment by famod** [comment]
Let's see what CI has to say. The rest is up to @glefloch and @quarkusio/devtools 

Thanks so far!

**Comment by geoand** [comment]
Looks like CI is not happy again :(

**Comment by snazy** [comment]
Issue 1 (Quarkus plugin code issue):
```
> Task :application:quarkusGenerateCode FAILED
Watching 43 directories to track changes
The configuration :application:quarkusProdBaseRuntimeClasspathConfiguration is both resolvable and consumable. This is considered a legacy configuration and it will eventually only be possible to be one of these.
The configuration :application:quarkusProdBaseRuntimeClasspathConfiguration is both consumable and declarable. This combination is incorrect, only one of these flags should be set.
```

Issue 2 (Quarkus test code issue):
```
[INFO] Running io.quarkus.gradle.InjectBeanFromTestConfigTest
$ /home/runner/work/quarkus/quarkus/integration-tests/gradle/./gradlew -Dmaven.repo.local=/home/runner/.m2/repository -Dorg.gradle.console=plain --stacktrace --info clean :application:test
Command: /home/runner/work/quarkus/quarkus/integration-tests/gradle/./gradlew -Dmaven.repo.local=/home/runner/.m2/repository -Dorg.gradle.console=plain --stacktrace --info clean :application:test ran from: /home/runner/work/quarkus/quarkus/integration-tests/gradle/target/classes/inject-bean-from-test-config failed with exit code: 1 and the following output:

FAILURE: Build failed with an exception.

* Where:
Build file '/home/runner/work/quarkus/quarkus/integration-tests/gradle/target/classes/inject-bean-from-test-config/library/build.gradle' line: 31

* What went wrong:
A problem occurred evaluating project ':library'.
> Could not set unknown property 'classifier' for task ':library:testJar' of type org.gradle.api.tasks.bundling.Jar.
```

**Comment by snazy** [comment]
Plus (it's a hard error in Gradle 8, was a warning in 7.x):
```
[WARN] [stderr] FAILURE: Build failed with an exception.
[WARN] [stderr] 
[WARN] [stderr] * What went wrong:
[WARN] [stderr] A problem was found with the configuration of task ':application:quarkusGenerateCode' (type 'QuarkusGenerateCode').
[WARN] [stderr]   - Gradle detected a problem with the following location: '/home/snazy/devel/quarkusio/quarkus/master/integration-tests/gradle/target/implementation-files12072569354363220854/common/build/libs/common.jar'.
[WARN] [stderr]     
[WARN] [stderr]     Reason: Task ':application:quarkusGenerateCode' uses this output of task ':common:jar' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.
[WARN] [stderr]     
[WARN] [stderr]     Possible solutions:
[WARN] [stderr]       1. Declare task ':common:jar' as an input of ':application:quarkusGenerateCode'.
[WARN] [stderr]       2. Declare an explicit dependency on ':common:jar' from ':application:quarkusGenerateCode' using Task#dependsOn.
[WARN] [stderr]       3. Declare an explicit dependency on ':common:jar' from ':application:quarkusGenerateCode' using Task#mustRunAfter.
```

**Comment by snazy** [comment]
Pushed another commit, it became a bit bigger than expected, but should fix the Gradle related issues in the code base. Most are integration test issues, and a few changes in the production Gradle plugin code.

File: devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/QuarkusPlugin.java [hunk 3]
```diff
@@ -210,23 +210,23 @@ private void registerTasks(Project project, QuarkusPluginExtension quarkusExt) {
                     TaskProvider<Task> testResourcesTask = tasks.named(JavaPlugin.PROCESS_TEST_RESOURCES_TASK_NAME);
 
                     quarkusGenerateCode.configure(task -> {
-                        task.dependsOn(resourcesTask);
-                        task.setCompileClasspath(
-                                project.getConfigurations().getByName(
-                                        ApplicationDeploymentClasspathBuilder.getBaseRuntimeConfigName(LaunchMode.NORMAL)));
+                        Configuration config = project.getConfigurations().getByName(
+                                ApplicationDeploymentClasspathBuilder.getBaseRuntimeConfigName(LaunchMode.NORMAL));
+                        task.dependsOn(resourcesTask, config);
```
Comment (snazy) [inline devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/QuarkusPlugin.java:215]: Add explicit dependsOn the configuration

File: devtools/gradle/gradle-model/src/main/java/io/quarkus/gradle/dependency/ApplicationDeploymentClasspathBuilder.java [hunk 4]
```diff
@@ -66,38 +66,50 @@ public static void initConfigurations(Project project) {
         final ConfigurationContainer configContainer = project.getConfigurations();
 
         // Custom configuration for dev mode
-        configContainer.create(ToolingUtils.DEV_MODE_CONFIGURATION_NAME)
-                .extendsFrom(configContainer.getByName(JavaPlugin.IMPLEMENTATION_CONFIGURATION_NAME));
+        configContainer.register(ToolingUtils.DEV_MODE_CONFIGURATION_NAME, config -> {
```
Comment (snazy) [inline devtools/gradle/gradle-model/src/main/java/io/quarkus/gradle/dependency/ApplicationDeploymentClasspathBuilder.java:69]: `.register` w/ config action (and also `.named`) are nicer (less code to run if the Quarkus Gradle tasks aren't needed in a build - aka "Gradle's configuration on demand")

File: integration-tests/gradle/src/main/resources/implementation-files/application-files/build.gradle [hunk 5]
```diff
@@ -0,0 +1,10 @@
+plugins {
+    id 'io.quarkus'
+}
+
+dependencies {
+    implementation files("../common/build/libs/common.jar")
```
Comment (snazy) [inline integration-tests/gradle/src/main/resources/implementation-files/application-files/build.gradle:6]: Dependencies on files w/o "proper" dependencies was always a bad idea, and is strictly enforced w/ Gradle 8.
I left the cross-project task dependencies here, although cross-project tasks should also never be used (this is a test, so ...). The correct way here is to use a `project(":common")` dependency, which was added to this test as well.
Note: If "raw" files "from somewhere else" are really needed, then either use project+configuration (in the same build) or "published" artifacts generated by other projects.

File: integration-tests/gradle/src/test/java/io/quarkus/gradle/BuildConfigurationTest.java [hunk 6]
```diff
@@ -1,108 +1,142 @@
...
 import java.nio.file.Path;
 
+import org.assertj.core.api.SoftAssertions;
+import org.assertj.core.api.junit.jupiter.InjectSoftAssertions;
+import org.assertj.core.api.junit.jupiter.SoftAssertionsExtension;
 import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
 
+@ExtendWith(SoftAssertionsExtension.class)
 public class BuildConfigurationTest extends QuarkusGradleWrapperTestBase {
```
Comment (snazy) [inline integration-tests/gradle/src/test/java/io/quarkus/gradle/BuildConfigurationTest.java:15]: Refactored the test to unify common test logic, to run less Gradle builds (faster test runtime) and also emit build output on failure.

File: integration-tests/gradle/src/test/java/io/quarkus/gradle/BuildResult.java [hunk 7]
```diff
@@ -57,6 +57,6 @@ public String getOutput() {
     }
 
     public static boolean isSuccessful(String result) {
-        return SUCCESS_OUTCOME.equals(result) || FROM_CACHE.equals(result);
+        return SUCCESS_OUTCOME.equals(result) || FROM_CACHE.equals(result) || UPTODATE_OUTCOME.equals(result);
```
Comment (snazy) [inline integration-tests/gradle/src/test/java/io/quarkus/gradle/BuildResult.java:60]: `UP_TO_DATE` is a valid "successful" outcome

File: integration-tests/gradle/src/test/java/io/quarkus/gradle/QuarkusGradleWrapperTestBase.java [hunk 8]
```diff
@@ -23,10 +25,32 @@ public class QuarkusGradleWrapperTestBase extends QuarkusGradleTestBase {
...
+        Runtime.getRuntime().addShutdownHook(new Thread(QuarkusGradleWrapperTestBase::killProcesses));
+    }
+
     protected void setupTestCommand() {
 
     }
 
+    @AfterAll
+    static void killProcesses() {
+        gradleProcesses.forEach(Process::destroy);
```
Comment (snazy) [inline integration-tests/gradle/src/test/java/io/quarkus/gradle/QuarkusGradleWrapperTestBase.java:39]: Ran into situations locally, when forked processes were still alive, so surefire forcefully terminated the test JVM. This after-all prevents the situation. (It doesn't kill a launched Gradle daemon, just the Gradle wrapper.)

File: integration-tests/gradle/src/test/java/io/quarkus/gradle/devmode/QuarkusDevGradleTestBase.java [hunk 9]
```diff
@@ -32,24 +33,21 @@ public void main() throws Exception {
         projectDir = getProjectDir();
         beforeQuarkusDev();
         ExecutorService executor = null;
-        final BuildResult[] buildResult = new BuildResult[1];
+        AtomicReference<BuildResult> buildResult = new AtomicReference<>();
```
Comment (snazy) [inline integration-tests/gradle/src/test/java/io/quarkus/gradle/devmode/QuarkusDevGradleTestBase.java:36]: Unlikely, but possible memory-issue with the array.

**Comment by maxandersen** [comment]
Does this change in any way which versions of grade quarkus users are able to use?

**Comment by snazy** [comment]
> Does this change in any way which versions of grade quarkus users are able to use?

No, there are no dependencies to "new" stuff available only in Gradle 8.

**Comment by gsmet** [comment]
Marked as draft for now. Please run CI in your fork first, then squash the commits into semantic commits then you can mark it as ready for review.

**Comment by gsmet** [comment]
@snazy can you mark it ready for review once CI is green for the Gradle tests in your fork?

**Comment by snazy** [comment]
Sure, is just took a while until all the checks [finished on my fork](https://github.com/snazy/quarkus/pull/171).

**Comment by snazy** [comment]
Haven't rebased it yet to keep the check results on my fork available. There are no conflicts though.

**Comment by famod** [comment]
FWIW, Gradle 8.0.1 is out: https://github.com/gradle/gradle/releases/tag/v8.0.1

**Comment by snazy** [comment]
Bumped to 8.0.1

**Comment by gsmet** [comment]
Rebased to get some CI fixes. Let's see what CI has to say...

**Review by aloubyansky — APPROVED** [review]
Thanks a lot @snazy Would you mind squashing your commits?

**Comment by snazy** [comment]
Sure, squashed & rebased

**Review by glefloch — APPROVED** [review]
Thanks a lot @snazy. This looks good to me!

**Pull request merged**